### ISSUE TITLE
Add other format?

### ISSUE THREAD

**Issue by ickc** [ickc/pantable#8]
Currently, pantable support to and from CSV.

Potentially, other table formats could be supported:

- [ ] `.xlsx`: this one will be useful but difficult:
    - Ideally, one would want to read/write `.xlsx` just like how pandoc read/write `.docx`. But there's seems no good cli to convert between `.docx` and `.xlsx` (to pass the `.docx` to pandoc).
    - So one might allow "markdown syntax" in `.xlsx`, which might works just like the current `.csv` but seems counterintuitive (people expect rich formatting in Excel).
    - Or find a tool to convert `.xlsx` to `.html` to pass the `.html` to pandoc (quite lossy though).
    - And then this' a question of what if people want to intermix rich text and markdown syntax (e.g. for LaTeX equation).
- [ ] HTML: if someone find writing the source table in HTML easier, say when it is a big table, but want to output to other formats as well
- [ ] YAML (table representing general YAML is not ideal, I'm thinking more like 2 column tables to show the key-value pairs)
- [ ] JSON (similar to above)

**Comment by ber532k** [comment]
xlsx-support without formating can be achieved quite easily using [pyexcel](https://github.com/pyexcel/pyexcel):

```
import pyexcel
raw_table_list = pyexcel.get_array(file_name="input.xlsx")
```

This might already take you quite a long way without too many changes to pantable.

**Comment by ickc** [comment]
I've considered something like that. But the hard part is to get "proper" xlsx support. Otherwise we're just "saving CSV in an excel container", so to speak.

e.g. looking at a macro in Excel for LaTeX (forgot the name but that's virtually the only game on town), people would expect one just to format the Excel in the usual way, and all translation happened 'automatically'.

Ideally, something like pandoc that parses docx but for xlsx would be appropriate. It is just that such tool isn't available. I've even considered converting the xlsx to HTML first and uses pandoc to translate it to the AST, but don't know how to do the first step "elegantly" and reliably.

**Comment by fralau** [comment]
1. The question of the csv separator and dialect is, in my view, slightly different because it involves very few changes in the code (the csv module provides the necessary arguments "out of the box"). It is also very important, because csv files come in all forms and shapes (see [RFC 480](https://tools.ietf.org/html/rfc4180)).

1. I really have a preference for the "Excel" dialect with semicolons (";") because *that is the format that Excel can import automatically*, without any manipulations (just double-click on the file). So working with that format saves a lot of time with manipulation, debugging, etc.

1. I am glad to see that [pyexcel](https://github.com/pyexcel/pyexcel) has gone such a long way with importing files (and notably arrays) from an xlrx file (natively), because a few years ago, I had develop my own logic with [xlrd](https://github.com/python-excel/xlrd), and that was not exactly fun. I works very well, though.

1. My suggestion is that if if you import .xlrx files, you do some reasoning about worksheets.

   - if no worksheet is specified, import data from the *first* worksheet
   - otherwise, accepte a 'worksheet' argument (either by name or by number). The use cases with the name or the number could be different, so I would provide both possibilities (and I suppose pyexcel should be agnostic anyway about this).

1. Also, note that csv files (and possibly Excel files) do not always start with the list of fields at the first line. This is due to the fact that some SQL export tools write the query on the first line of the export file, by default. Having to  edit csv files manually every time can be annoying, so an argument saying e.g. "first_line = 2" could be useful.

**Comment by gepcel** [comment]
Maybe use pandas? pandas can read data from xlsx, json, html, and [many others](http://pandas.pydata.org/pandas-docs/stable/user_guide/io.html).

There's a temp to implement this in #37 . Right now, I'm using it myself, works fine.

**Comment by gepcel** [comment]
Pantable works very well with csv. Just for xlsx, json, html, even from an online webpage and many others, pandas may be a solution. Just my opinion, for me, csv and my own implementation of xlsx are enough for me.

**Comment by gepcel** [comment]
Ha, maybe I wasn't express myself clearly, the way pantale delt with csv is very good. I wasn't suggesting switching to pandas. And I wan't pushing my opition or my pull request either.

As for xlsx, I wasn't thinking as much as you do. For me, I actually don't need any advantage of xlsx, so what I did [here](https://github.com/ickc/pantable/pull/37) works for me.

As for why xlsx at all,  from where I'm working, nobody around uses csv, nobody knows pandoc, even nobody knows markdown, the whole world is full of docx and xlsx. (^_^ I'm lonely.). The format I import all the data from and export to is xlsx, this format is what I'm given with and what I will share with. But I honestly don't need any more markups from xlsx, just like bold, or italic or some equation, I can use raw markdown in xlsx.

So basically, for me, xlsx is just a  csv file with no format markups so that others can be opened directly with excel.

Following the way you were thinking of, pandas wasn't a good way to read the *table*, but just a convenient way to read *data* (at least it gets the data right). I like the 'adventurous' idea, so maybe for some formats, you can implement them one by one, and for others, make theme under alpha only.

Thank you for making this very useful soft.

**Comment by ickc** [comment]
Excel open CSV just fine. So I've been trying to save as CSV rather than xlsx if I'm not using "advanced" Excel features. But then I recently also run into another problem that I end up saving in xlsx even when all the cells are plain text, because I need the split and freeze functionality in order to navigate a big table effectively.

So I guess there's valid argument to save in xlsx even when cells are plain texts.

I'm more leaning toward writing another filter to preprocess the files as csv first. e.g. a filter using pandas exclusively to do just that, leaving options for another filter using another backend. Naming this would be interesting, combining pandoc and pandas? pandacs?
